### PR TITLE
fix: apikey plan selection

### PR DESCRIPTION
**Pull request by jhaeyaert**
## Issue

https://github.com/gravitee-io/issues/issues/8452

## Description

This PR fixes the plan selection workflow to go back to the previous one. This issue has been introduced while fixing another security issue on JWT/Oaut2 plan selection.

When defining an ApiKey plan and a Keyless plan, the new implemented workflow tried to identify the plan by filtering on the associated subscription.
When an api key is provided in the request but no valid subscription is found the keyless plan is executed -> this wasn't the case previously.

This PR go back to the previous behavior which is to let the ApiKey plan reject the call when an apikey is provided in the request. In the same time it keeps the benefit of the security fix on JWT/Oauth2 plan.

🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.z6.web.core.windows.net/8452-fix-plan-selection/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-fufqypbwtd.chromatic.com)

### CODE REVIEW

File: gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/handlers/api/security/ApiKeyPlanBasedAuthenticationHandler.java [hunk 1]
```diff
@@ -40,11 +40,19 @@ public ApiKeyPlanBasedAuthenticationHandler(AuthenticationHandler handler, Plan
      * {@inheritDoc}
      */
     @Override
-    protected boolean canHandleSubscription(AuthenticationContext authenticationContext) {
+    protected boolean preCheckSubscription(AuthenticationContext authenticationContext) {
```
Comment (guillaumelamirand) [inline gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/handlers/api/security/ApiKeyPlanBasedAuthenticationHandler.java:43]: I don't understand the naming as there is not check or post check Subscription

File: gravitee-apim-gateway/gravitee-apim-gateway-security/gravitee-apim-gateway-security-core/src/main/java/io/gravitee/gateway/security/core/DefaultAuthenticationHandlerSelector.java [hunk 2]
```diff
@@ -33,11 +35,23 @@ public DefaultAuthenticationHandlerSelector(AuthenticationHandlerManager authent
...
-        final SimpleAuthenticationContext context = new SimpleAuthenticationContext(executionContext);
+        final SimpleAuthenticationContext authenticationContext = new SimpleAuthenticationContext(executionContext);
+        final List<AuthenticationHandler> authenticationHandlers = authenticationHandlerManager.getAuthenticationHandlers();
+
+        for (int i = 0; i < authenticationHandlers.size(); i++) {
+            final AuthenticationHandler authenticationHandler = authenticationHandlers.get(i);
+            authenticationContext.setInternalAttribute(ATTR_INTERNAL_LAST_SECURITY_HANDLER_SUPPORTING_SAME_TOKEN_TYPE, true);
+
+            if (i < authenticationHandlers.size() - 1) {
+                // Check whether the next authentication handler works on the same type of token or not.
```
Comment (guillaumelamirand) [inline gravitee-apim-gateway/gravitee-apim-gateway-security/gravitee-apim-gateway-security-core/src/main/java/io/gravitee/gateway/security/core/DefaultAuthenticationHandlerSelector.java:46]: Are they sorted by Type ? Shouldn't we test if one of next oneS support the same type?

File: gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/handlers/api/security/ApiKeyPlanBasedAuthenticationHandler.java (hunk 1 above)
Comment (jhaeyaert) [reply gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/handlers/api/security/ApiKeyPlanBasedAuthenticationHandler.java:43]: The naming is definitely not the best. It is just something with discussed with other team members and appears to be clearer than `canHandleSubscription`.

The preCheckSubscription is used in a particular case (jwt) and can be skipped under certain circumstances.

File: gravitee-apim-gateway/gravitee-apim-gateway-security/gravitee-apim-gateway-security-core/src/main/java/io/gravitee/gateway/security/core/DefaultAuthenticationHandlerSelector.java (hunk 2 above)
Comment (jhaeyaert) [reply gravitee-apim-gateway/gravitee-apim-gateway-security/gravitee-apim-gateway-security-core/src/main/java/io/gravitee/gateway/security/core/DefaultAuthenticationHandlerSelector.java:46]: They are sorted by order(). We also discuss this point and trying to change that order by reordering or grouping by type has been quickly abandoned due to possible bad side effects.

**Review by marcambier — APPROVED** [review]

**Pull request merged**